Prepare ec2_vpc_vgw_* modules for promotion

mandar242 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Depends-On: https://github.com/ansible-collections/amazon.aws/pull/2331
Prepare ec2_vpc_vgw_* modules for promotion

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ec2_vpc_vgw
ec2_vpc_vgw_info

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -211,26 +217,26 @@ def wait_for_status(client, module, vpn_gateway_id, status):
                 time.sleep(polling_increment_secs)
-        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-            module.fail_json_aws(e, msg="Failure while waiting for status update")
+        except AnsibleEC2Error as e:
```
Can't you define a custom waiter instead?

mandar242 replied:
added custom waiter https://github.com/ansible-collections/amazon.aws/pull/2331

alinabuzachis commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -88,15 +83,15 @@
       tags:
         description:
-          - A list of dictionaries representing the tags attached to the VGW.
+          - A list of dictionaries representing the tags attached to the virtual private gateway..
```
```suggestion
          - A list of dictionaries representing the tags attached to the virtual private gateway.
```

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
Can you please update description? 

description:
  - Creates AWS VPN Virtual Gateways.
  - Deletes AWS VPN Virtual Gateways.
  - Attaches Virtual Gateways to VPCs.
  - Detaches Virtual Gateways from VPCs.

Same for 

notes:
  - Support for O(purge_tags) was added in release 4.0.0.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -88,15 +83,15 @@
         description:
-          - A list of dictionaries representing the tags attached to the VGW.
+          - A list of dictionaries representing the tags attached to the virtual private gateway..
           - Represents the same details as I(resource_tags).
```
```suggestion
          - Represents the same details as RV(resource_tags).
```

alinabuzachis commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -106,28 +101,28 @@
       resource_tags:
         description:
           - A dictionary representing the tags attached to the VGW.
           - Represents the same details as I(tags).
```
```suggestion
          - Represents the same details as RV(tags).
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +205,54 @@ def get_vgw_info(vgws):
 
 
-def wait_for_status(client, module, vpn_gateway_id, status):
+def wait_for_status(client, module, vpn_gateway_id, desired_status):
```
could you please add type hint here?

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +205,54 @@ def get_vgw_info(vgws):
-    result = response
-    return status_achieved, result
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +205,54 @@ def get_vgw_info(vgws):
+        # response = attach_vpc_vpn_gateway(client, vpn_gateway_id, params["VpcId"])
+        response = attach_vpc_vpn_gateway(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

GomathiselviS replied:
Shouldn't these errors be handled in the amazon.aws.ec2 utility functions?

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -244,13 +266,15 @@ def detach_vgw(client, module, vpn_gateway_id, vpc_id=None):
-        module.fail_json_aws(e, "Failed to detach gateway")
+        response = detach_vpc_vpn_gateway(client, **detach_params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,29 +294,29 @@ def create_vgw(client, module):
-    ) as e:  # pylint: disable=duplicate-except
-        module.fail_json_aws(e, msg="Failed to create gateway")
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,29 +294,29 @@ def create_vgw(client, module):
-        module.fail_json_aws(e, msg="Failed to delete gateway")
+        delete_vpc_vpn_gateway(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -301,13 +325,14 @@ def delete_vgw(client, module, vpn_gateway_id):
-            module.fail_json_aws(e, msg="Failed to describe VPC")
+            response = describe_vpcs(client, **params)
+        except AnsibleEC2Error as e:
+            module.fail_json_aws(e)
```
```suggestion
            module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,11 +350,11 @@ def find_vgw(client, module, vpn_gateway_id=None):
-        module.fail_json_aws(e, msg="Failed to describe gateway using filters")
+        response = describe_vpc_vpn_gateways(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -150,20 +145,22 @@ def get_virtual_gateway_info(virtual_gateway):
-        module.fail_json_aws(e, msg="Failed to list gateways")
+        all_virtual_gateways = describe_vpc_vpn_gateways(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
```
```suggestion
        module.fail_json_aws_error(e)
```

abikouo reviewed (changes requested):
Please update the documentation of the `ec2_vpc_vgw` module, the `state` argument has values `C(present)` and `C(absent)`, should be `V(present)` and `V(absent)` instead.
You also need to update the `ec2_vpc_vgw` integration test case: add uppercase to tasks names and use FQCN for module.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +205,54 @@ def get_vgw_info(vgws):
-        )(VpnGatewayId=vpn_gateway_id, VpcId=params["VpcId"])
-    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-        module.fail_json_aws(e, msg="Failed to attach VPC")
+        # response = attach_vpc_vpn_gateway(client, vpn_gateway_id, params["VpcId"])
```
Is this commented line needed?

mandar242 replied:
removed. left over from testing

alinabuzachis reviewed (commented):
Can you also check why some tasks are commented out in the integration tests https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_vgw/tasks/tags.yml and make sure that the module is fully tested, adding more tests if necessary?

GomathiselviS commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -89,48 +87,49 @@
+      description: The ID of the virtual private gateway.
       type: str
       returned: success
       example: "vgw-0123456789abcdef0"
```
```suggestion
      sample: "vgw-0123456789abcdef0"
```

GomathiselviS replied:
This has to be changed for all the parameters to maintain consistency

GomathiselviS commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -32,54 +31,50 @@
+        description: The ID of the virtual private gateway.
         type: str
         returned: success
         example: "vgw-0123456789abcdef0"
```
```suggestion
        sample: "vgw-0123456789abcdef0"
```

mandar242 commented:
> Can you also check why some tasks are commented out in the integration tests https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_vgw/tasks/tags.yml and make sure that the module is fully tested, adding more tests if necessary?

ec2_vpc_vgw did not had check_mode support. Added check_mode support and enabled tests, added more tests.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+        "VpcId": module.params.get("vpc_id"),
+        "VpnGatewayIds": module.params.get("vpn_gateway_id"),
+        "Tags": module.params.get("tags") or {},
+        "PurgeTags": module.params.get("purge_tags", False),
```
I do not think "PurgeTags" must be passed to the API call.

mandar242 replied:
removed.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
-    params = dict()
-    params["VpcId"] = module.params.get("vpc_id")
+
+def attach_vgw(client, module: AnsibleAWSModule, vpn_gateway_id: str) -> Any:
```
Is check_mode handled for this operation?

mandar242 replied:
fixed

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -240,17 +265,13 @@ def attach_vgw(client, module, vpn_gateway_id):
-    params = dict()
-    params["VpcId"] = module.params.get("vpc_id")
+def detach_vgw(client, module: AnsibleAWSModule, vpn_gateway_id: str, vpc_id: Optional[str] = None) -> Any:
+    vpc_id = vpc_id or module.params.get("vpc_id")
```
I check_mode handled for this operation?

mandar242 replied:
fixed

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,50 +291,48 @@ def create_vgw(client, module):
+        except AnsibleEC2Error as e:
+            module.fail_json_aws_error(e)
 
     result = response
     return result
```
Ad I would initialise response before.

```suggestion
    return response
```

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -240,17 +265,13 @@ def attach_vgw(client, module, vpn_gateway_id):
     return result
```
Same

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
-    status_achieved = False
+    try:
+        wait_for_resource_state(client, module, "vpn_gateway_exists", VpnGatewayIds=vpn_gateway_id)
+        if desired_status == "attached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_attached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
+        elif desired_status == "detached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_detached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
```
```suggestion
        if desired_status not in ("attached", "detached"):
              module.fail_json(msg=f"Unsupported status: {desired_status}")
        wait_for_resource_state(
                client,
                module,
                f"vpn_gateway_{desired_status}",
                VpnGatewayIds=vpn_gateway_id,
                delay=polling_increment_secs,
                max_attempts=max_retries,
            )
       
```

mandar242 replied:
changing VpnGatewayIds=vpn_gateway_id to VpnGatewayIds=[vpn_gateway_id] causes failure with
```
"msg": "An exception happened while trying to wait for 'vpn_gateway_attached': Parameter validation failed:\nInvalid type for parameter VpnGatewayIds[0], value: ['vgw-0c3df60baf0acc49f'], type: <class 'list'>, valid types: <class 'str'>",
    "resource_actions": [
        "ec2:DescribeVpnGateways",
        "ec2:CreateVpnGateway",
        "ec2:AttachVpnGateway"
    ]
```
even though API states otherwise it to be a list of str, weird.

alinabuzachis replied:
Same as below.

mandar242 replied:
Hi @alinabuzachis , fixed. It was indeed being passed as List[Str] at the API call.
Only thing was that it was handled to be passed as List[Str] earlier in the function call tree.
Updated the code to ensure it is passed as List[Str] before calling module_utils/ec2 functions.
https://github.com/ansible-collections/community.aws/pull/2171/commits/27b07e056bccfd390b8c1425acfdaa6ac58a44af

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,50 +291,48 @@ def create_vgw(client, module):
 
-def find_vgw(client, module, vpn_gateway_id=None):
+def find_vgw(
+    client, module: AnsibleAWSModule, vpn_gateway_id: Optional[Union[str, List[str]]] = None
```
```suggestion
    client, module: AnsibleAWSModule, vpn_gateway_id: Optional[str] = None
```

mandar242 replied:
Hi @alinabuzachis @abikouo  , fixed. It was indeed being passed as List[Str] at the API call.
Only thing was that it was handled to be passed as List[Str] earlier in the function call tree.
Updated the code to ensure it is passed as List[Str] before calling module_utils/ec2 functions.
https://github.com/ansible-collections/community.aws/pull/2171/commits/27b07e056bccfd390b8c1425acfdaa6ac58a44af

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,50 +291,48 @@ def create_vgw(client, module):
+) -> List[Dict[str, Any]]:
     params = dict()
     if vpn_gateway_id:
         params["VpnGatewayIds"] = vpn_gateway_id
```
```suggestion
        params["VpnGatewayIds"] = [vpn_gateway_id]
```

mandar242 replied:
not sure what happens with the API but it fails with
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Invalid type for parameter VpnGatewayIds[0], value: ['vgw-0849b49ec813d18ee'], type: <class 'list'>, valid types: <class 'str'>
```

alinabuzachis replied:
Same as below.

mandar242 replied:
Hi @alinabuzachis , fixed. It was indeed being passed as List[Str] at the API call.
Only thing was that it was handled to be passed as List[Str] earlier in the function call tree.
Updated the code to ensure it is passed as List[Str] before calling module_utils/ec2 functions.
https://github.com/ansible-collections/community.aws/pull/2171/commits/27b07e056bccfd390b8c1425acfdaa6ac58a44af

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
-    # found and we will not create another vgw.
+def ensure_vgw_present(client, module: AnsibleAWSModule) -> Tuple[bool, Dict[str, Any]]:
+    changed = False
+    params = {
+        "Name": module.params.get("name"),
+        "VpcId": module.params.get("vpc_id"),
+        "VpnGatewayIds": module.params.get("vpn_gateway_id"),
+        "Tags": module.params.get("tags") or {},
+        "PurgeTags": module.params.get("purge_tags", False),
```
Also, I don't understand why params is needed, since you only used params["VpcId"] in handle_existing_vgw() and create_new_gw(), which can be accessed from module.params, or am I wrong?

mandar242 replied:
right. removed.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
+def wait_for_status(
+    client, module: AnsibleAWSModule, vpn_gateway_id: Union[str, List[str]], desired_status: str
+) -> Tuple[bool, Any]:
     polling_increment_secs = 15
     max_retries = module.params.get("wait_timeout") // polling_increment_secs
```
I think these two are not needed anymore here.

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
 
-def wait_for_status(client, module, vpn_gateway_id, status):
+def wait_for_status(
+    client, module: AnsibleAWSModule, vpn_gateway_id: Union[str, List[str]], desired_status: str
```
```suggestion
    client, module: AnsibleAWSModule, vpn_gateway_id: str, desired_status: str
```

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
     max_retries = module.params.get("wait_timeout") // polling_increment_secs
-    status_achieved = False
+    try:
+        wait_for_resource_state(client, module, "vpn_gateway_exists", VpnGatewayIds=vpn_gateway_id)
```
```suggestion
        wait_for_resource_state(client, module, "vpn_gateway_exists", VpnGatewayIds=[vpn_gateway_id])
```

mandar242 replied:
not sure what happens with the API but it fails with
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Invalid type for parameter VpnGatewayIds[0], value: ['vgw-0849b49ec813d18ee'], type: <class 'list'>, valid types: <class 'str'>
```

alinabuzachis replied:
According to the API, VpnGatewayIds must be a list of strings.
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2/client/describe_vpn_gateways.html
The error you get may be because you are not passing a list of strings. You need to verify that a list of strings is always being passed.

mandar242 replied:
Hi @alinabuzachis , fixed. It was indeed being passed as List[Str] at the API call.
Only thing was that it was handled to be passed as List[Str] earlier in the function call tree.
Updated the code to ensure it is passed as List[Str] before calling module_utils/ec2 functions.
https://github.com/ansible-collections/community.aws/pull/2171/commits/27b07e056bccfd390b8c1425acfdaa6ac58a44af

alinabuzachis commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+                attach_vgw(client, module, vpn_gateway_id)
+                changed = True
+        else:
+            attach_vgw(client, module, vpn_gateway_id)
```
How is check_mode handled here?

mandar242 replied:
added check_mode handling to function `attach_vgw_to_vpc()` code

GomathiselviS commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,50 +291,48 @@ def create_vgw(client, module):
-        botocore.exceptions.BotoCoreError,
-    ) as e:  # pylint: disable=duplicate-except
-        module.fail_json_aws(e, msg="Failed to create gateway")
+    except AnsibleEC2Error as e:
```
Should this be handled here? Isn't this covered by https://github.com/ansible-collections/amazon.aws/pull/2331/files#diff-11d6a927433f45aa84de2f54fa0adf33e8094ebb417c75ea476ffe61b4e4f587R652 ?

mandar242 replied:
removed.

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -176,7 +190,7 @@ def found(response_code, catch_extra_error_codes=None):
 
 
-def get_vgw_info(vgws):
+def get_vgw_info(vgws: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
```
```suggestion
def get_vgw_info(vgw: Dict[str, Any]) -> Dict[str, Any]:
```
The function should be renamed to `format_vpn_gateway()` and accept a `dict` as input, not a list of dicts. We should not loop on vgws but threat the single element.
We should check whether the input is not an empty dict to deal with the specific case of `check_mode`

mandar242 replied:
fixed

GomathiselviS commented on plugins/modules/ec2_vpc_vgw_info.py:
```diff
@@ -150,20 +145,22 @@ def get_virtual_gateway_info(virtual_gateway):
-        module.fail_json_aws(e, msg="Failed to list gateways")
+        all_virtual_gateways = describe_vpn_gateways(client, **params)
+    except AnsibleEC2Error as e:
+        module.fail_json_aws_error(e)
```
error handling is done here - https://github.com/ansible-collections/amazon.aws/pull/2331/files#diff-11d6a927433f45aa84de2f54fa0adf33e8094ebb417c75ea476ffe61b4e4f587R652

mandar242 replied:
removed.

GomathiselviS commented on tests/integration/targets/ec2_vpc_vgw/tasks/main.yml:
```diff
@@ -23,66 +21,116 @@
 
-    - name: use set fact for vgw ids
+    - name: Use set fact for vgw ids
       set_fact:
```
```suggestion
      ansible.builtin.set_fact:
```

mandar242 replied:
fixed

GomathiselviS commented on tests/integration/targets/ec2_vpc_vgw/tasks/main.yml:
```diff
@@ -23,66 +21,116 @@
 
-    - name: get VPC VGW facts by ID (CHECK)
+    - name: Get VPC VGW facts by ID (CHECK)
       ec2_vpc_vgw_info:
```
Use FQCN

mandar242 replied:
fixed

GomathiselviS commented on tests/integration/targets/ec2_vpc_vgw/tasks/main.yml:
```diff
@@ -150,37 +198,52 @@
 
-    - name: get all VGWs
+    - name: Get all VGWs
       ec2_vpc_vgw_info:
```
Use FQCN

mandar242 replied:
fixed

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+    )
+
+    if existing_vgw:
+        changed |= handle_existing_vgw(client, module, existing_vgw[0], params)
```
```suggestion
        changed |= handle_existing_vgw(client, module, existing_vgw[0], params)
        vgw = find_vgw(client, module, [existing_vgw[0]["VpnGatewayId"]])
```

mandar242 replied:
this breaks the code logic

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+    else:
+        changed |= create_new_vgw(client, module, params)
+
+    vgw = find_vgw(client, module, [existing_vgw[0]["VpnGatewayId"]]) if existing_vgw else find_vgw(client, module)
+    return changed, get_vgw_info(vgw)
```
```suggestion
    return changed, get_vgw_info(vgw)
```

mandar242 replied:
this breaks the code logic

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+        else:
+            attach_vgw(client, module, vpn_gateway_id)
+            changed = True
+    else:
+        if current_vpc_attachments and current_vpc_attachments[0]["State"] == "attached":
+            if module.check_mode:
+                return True
+            detach_vgw(client, module, vpn_gateway_id, current_vpc_attachments[0]["VpcId"])
+            changed = True
```
```suggestion
```
The VPC should not be detached the Gateway, here nothing has to be done has the VPC was not specified and state is set to present. The VPC is detached when state is absent and VPC id is provided

mandar242 replied:
I'm not sure if that's correct. The current integration tests indicate that if
`state = present` and `vpc_id` is `not specified` then the `vgw` will be `detached` from the `vpcs`.

These tests are pre-existing, so module behavior is that before I started with refactoring
```
    - name: Detach vpn gateway
      community.aws.ec2_vpc_vgw:
        state: present
        name: "{{ vgw_name }}"
      register: vgw

    - name: Assert results
      ansible.builtin.assert:
        that:
          - vgw.changed
          - not vgw.vgw.vpc_id
   ```

similarly when `state = absent` then `vgw` will be `deleted` (by first detaching it from vpc)
```
    - name: Delete vpn gateway
      community.aws.ec2_vpc_vgw:
        state: absent
        vpn_gateway_id: '{{ vgw.vgw.id | default(vgw_id) }}'
      ignore_errors: true
```

abikouo commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
+    current_vpc_attachments = existing_vgw["VpcAttachments"]
     if params["VpcId"]:
-        vpc = find_vpc(client, module)
+        if current_vpc_attachments and current_vpc_attachments[0]["State"] == "attached":
+            current_vpc_id = current_vpc_attachments[0]["VpcId"]
+            if current_vpc_id != params["VpcId"]:
+                if module.check_mode:
+                    return True
+                detach_vgw(client, module, vpn_gateway_id, current_vpc_id)
+                get_waiter(client, "vpn_gateway_detached").wait(VpnGatewayIds=[vpn_gateway_id])
+                attach_vgw(client, module, vpn_gateway_id)
+                changed = True
+        else:
+            attach_vgw(client, module, vpn_gateway_id)
+            changed = True
```
```suggestion
        # check first if the VPC is already attached to the VPN
        if any(params["VpcId"] == attachment["VpcId"] for attachment in current_vpc_attachments):
            return changed
        # Now attach the VPC to the Gateway
        if module.check_mode:
            return changed
        attach_vgw(client, module, vpn_gateway_id)
        changed = True
```
The Gateway can be attached to multiple VPC, we should not always detach one attached before attaching another.

mandar242 replied:
as far as I understand that was the existing functionality, so did not wanted to change anything that would introduce breaking changes by accident. For now I think it would be okay to move the module to a.aws with same functionality and this issue could be tackled later
@alinabuzachis @abikouo

abikouo reviewed (changes requested):
Lot of things to change here. Generally the `ec2_vpc_vgw` could be simplified and you need to ensure that the module is working as expected.

`state==present` -> Create the Gateway if it does not exist, attach the VPC to the gateway, update tags if they have changed
`state==absent, vpc_id attached to the Gateway` -> Detach the VPC to the Gateway, but **do not delete the Gateway**
`state==absent` -> Detach all VPC attached to the Gateway, and **delete the Gateway**

mandar242 commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -270,50 +291,48 @@ def create_vgw(client, module):
+) -> List[Dict[str, Any]]:
     params = dict()
     if vpn_gateway_id:
         params["VpnGatewayIds"] = vpn_gateway_id
```
not sure what happens with the API but it fails with
```
    "msg": "Failed to describe vpn gateways: Parameter validation failed:\nInvalid type for parameter VpnGatewayIds[0], value: ['vgw-0b7cbbd0c610a9e10'], type: <class 'list'>, valid types: <class 'str'>",
    "resource_actions": [
        "ec2:CreateVpnGateway",
        "ec2:AttachVpnGateway",
        "ec2:DescribeVpnGateways"
    ]
```

mandar242 replied:
Hi @alinabuzachis , fixed. It was indeed being passed as List[Str] at the API call.
Only thing was that it was handled to be passed as List[Str] earlier in the function call tree.
Updated the code to ensure it is passed as List[Str] before calling module_utils/ec2 functions.
https://github.com/ansible-collections/community.aws/commit/27b07e056bccfd390b8c1425acfdaa6ac58a44af

mandar242 commented:
> Lot of things to change here. Generally the `ec2_vpc_vgw` could be simplified and you need to ensure that the module is working as expected.
> 
> `state==present` -> Create the Gateway if it does not exist, attach the VPC to the gateway, update tags if they have changed `state==absent, vpc_id attached to the Gateway` -> Detach the VPC to the Gateway, but **do not delete the Gateway** `state==absent` -> Detach all VPC attached to the Gateway, and **delete the Gateway**

I'm not sure on 
```
state==absent, vpc_id attached to the Gateway` -> Detach the VPC to the Gateway, but **do not delete the` 
```

because the way current tests are set up (existing before I started with refactoring), to detach the vpn gw, state is used as present
```
    - name: Detach vpn gateway
      community.aws.ec2_vpc_vgw:
        state: present
        name: "{{ vgw_name }}"
      register: vgw
```

changing the module behavior will likely cause breaking change
@abikouo

mandar242 commented on changelogs/fragments/refactor-ec2_vpc_vgw-modules.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - ec2_vpc_vgw - Refactor module to use shared code from ``amazon.aws.plugins.module_utils.ec2`` and update ``RETURN`` block (https://github.com/ansible-collections/community.aws/pull/2127).
+  - ec2_vpc_vgw_info - Refactor module to use shared code from ``amazon.aws.plugins.module_utils.ec2`` and update ``RETURN`` block (https://github.com/ansible-collections/community.aws/pull/2127).
```
fixed

mandar242 commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -9,10 +9,10 @@
+  - Detaches Virtual Gateways from VPCs.
 options:
   state:
     description:
```
fixed

mandar242 commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -431,6 +447,8 @@ def ensure_vgw_absent(client, module):
+            if module.check_mode:
+                return True, ""
             existing_vgw = existing_vgw_with_id
             if existing_vgw[0]["VpcAttachments"] != [] and existing_vgw[0]["VpcAttachments"][0]["State"] == "attached":
```
based on existing functionality, and judging from existing integration tests, I believe that if `state = absent` then VGW should be deleted.
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_vgw/tasks/main.yml

mandar242 commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -485,6 +505,8 @@ def ensure_vgw_absent(client, module):
                         changed = True
```
based on existing functionality, and judging from existing integration tests, I believe that if state = absent then VGW should be deleted.
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_vgw/tasks/main.yml

to detach, user needs can https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/ec2_vpc_vgw/tasks/main.yml#L165-L171
```
    - name: detach vpn gateway
      ec2_vpc_vgw:
        state: present
        name: "{{ vgw_name }}"
      register: vgw
```

mandar242 commented on plugins/modules/ec2_vpc_vgw.py:
```diff
@@ -485,6 +505,8 @@ def ensure_vgw_absent(client, module):
+                    if module.check_mode:
+                        return True, ""
                     # attempt to detach any attached vpcs
                     vpc_to_detach = existing_vgw[0]["VpcAttachments"][0]["VpcId"]
                     detach_vgw(client, module, vpn_gateway_id, vpc_to_detach)
```
fixed.

alinabuzachis commented:
recheck

alinabuzachis commented:
recheck

alinabuzachis commented:
recheck

alinabuzachis commented:
@mandar242 there are still other linting issues.
